### PR TITLE
Seeding in all samplers, xi and mu and islocal fully available

### PR DESCRIPTION
**Pull request by KrainskiL**
I think that `abcd_sampler.jl` and other samplers should stay as they are - user can use all-in-one generation with `abcd_sampler` or step by step, more customized sampling with `com_sampler`, `deg_sampler` and `graph_sampler`. Merging both approaches may create excessively complex and unclear configuration file.

### CODE REVIEW

**Comment by bkamins** [comment]
@pawel-pralat - are you OK with this?

File: README.md [hunk 1]
```diff
@@ -67,6 +70,8 @@ $ julia abcd_sampler.jl example_config.toml
 [ Info: For the syntax of config_filename see example_config.toml file
 [ Info: Expected value of degree: 8.327743727955891
 [ Info: Expected value of community size: 156.5613820733916
+$ shasum -a 256 edge.dat
```
Comment (bkamins) [inline README.md]: add above that it will work under Linux pls.

File: utils/abcd_sampler.jl [hunk 2]
```diff
@@ -9,6 +9,12 @@ filename = ARGS[1]
 conf = Pkg.TOML.parsefile(filename)
 isempty(conf["seed"]) || Random.seed!(parse(Int, conf["seed"]))
 
+μ = haskey(conf,"mu") ? parse(Float64, conf["mu"]) : nothing
+ξ = haskey(conf,"xi") ? parse(Float64, conf["xi"]) : nothing
```
Comment (bkamins) [inline utils/abcd_sampler.jl]:
```suggestion
μ = haskey(conf, "mu") ? parse(Float64, conf["mu"]) : nothing
ξ = haskey(conf, "xi") ? parse(Float64, conf["xi"]) : nothing
```

File: utils/abcd_sampler.jl [hunk 3]
```diff
@@ -27,9 +33,10 @@ c_max_iter = parse(Int, conf["c_max_iter"])
 coms = ABCDGraphGenerator.sample_communities(τ₂, c_min, c_max, n, c_max_iter)
 open(io -> foreach(d -> println(io, d), coms), conf["communitysizesfile"], "w")
 
-ξ = parse(Float64, conf["xi"])
 isCL = parse(Bool, conf["isCL"])
-p = ABCDGraphGenerator.ABCDParams(degs, coms, nothing, ξ, isCL, false)
+islocal = haskey(conf,"islocal") ? parse(Bool, conf["islocal"]) : false
```
Comment (bkamins) [inline utils/abcd_sampler.jl]:
```suggestion
islocal = haskey(conf, "islocal") ? parse(Bool, conf["islocal"]) : false
```

File: utils/abcd_sampler.jl [hunk 4]
```diff
@@ -27,9 +33,10 @@ c_max_iter = parse(Int, conf["c_max_iter"])
...
+
+p = ABCDGraphGenerator.ABCDParams(degs, coms, μ, ξ, isCL, islocal)
```
Comment (bkamins) [inline utils/abcd_sampler.jl:39]: so here we allow passing `islocal=true` when ξ is passed - does the code throw an error later?

File: utils/deg_sampler.jl [hunk 5]
```diff
@@ -1,17 +1,19 @@
...
 n = parse(Int, ARGS[5])
 max_iter = parse(Int, ARGS[6])
+length(ARGS) == 7 && Random.seed!(parse(Int, ARGS[7]))
 
 @info "Expected value of degree: $(ABCDGraphGenerator.get_ev(τ₁, d_min, d_max))"
 
 degs = ABCDGraphGenerator.sample_degrees(τ₁, d_min, d_max, n, max_iter)
 
-open(io -> foreach(d -> println(io, d), degs), filename, "w")
+open(io -> foreach(d -> println(io, d), degs), filename, "w")
```
Comment (bkamins) [inline utils/deg_sampler.jl]: newline is missing

File: utils/graph_sampler.jl [hunk 6]
```diff
@@ -1,19 +1,30 @@
...
-μ = parse(Float64, ARGS[5])
-isCL = parse(Bool, ARGS[6])
-islocal = parse(Bool, ARGS[7])
+muxi = ARGS[5]
+fraction = parse(Float64, ARGS[6])
+isCL = parse(Bool, ARGS[7])
+islocal = parse(Bool, ARGS[8])
+
+muxi in ["mu","xi"] || throw(ArgumentError("only mu or xi names are allowed for"))
+μ = ξ = nothing
```
Comment (bkamins) [inline utils/graph_sampler.jl]:
```suggestion
μ, ξ = nothing, nothing
```

File: utils/graph_sampler.jl [hunk 7]
```diff
@@ -1,19 +1,30 @@
 using ABCDGraphGenerator
+using Random
 
-@info "Usage: julia graph_sampler.jl networkfile communityfile degreefile communitysizesfile μ isCL islocal"
-@info "Example: julia graph_sampler.jl network.dat community.dat degrees.dat community_sizes.dat 0.2 true true"
+@info "Usage: julia graph_sampler.jl networkfile communityfile degreefile communitysizesfile mu|xi fraction isCL islocal [seed]"
```
Comment (bkamins) [inline utils/graph_sampler.jl:4]: seed processing seems to be missing here

File: utils/graph_sampler.jl [hunk 8]
```diff
@@ -1,19 +1,30 @@
...
+    μ = fraction
+else
+    ξ = fraction
+end
+length(ARGS) == 9 &&  Random.seed!(parse(Int, ARGS[9]))
 
 p = ABCDGraphGenerator.ABCDParams(parse.(Int, readlines(degreefile)),
                                   parse.(Int, readlines(communitysizesfile)),
-                                  μ, nothing, isCL, islocal)
+                                  μ, ξ, isCL, islocal)
```
Comment (bkamins) [inline utils/graph_sampler.jl:27]: so again - we allow specifying ξ and passing islocal=true. I assume code throws an error later?

File: utils/abcd_sampler.jl (hunk 4 above)
Comment (KrainskiL) [reply utils/abcd_sampler.jl:39]: Yes, it's handled in `src/graph_sampler.jl` - add additional error in above file to catch it sooner?

File: utils/graph_sampler.jl (hunk 8 above)
Comment (KrainskiL) [reply utils/graph_sampler.jl:27]: Yes, it's handled in called code - again it can be handled sooner in above file if needed.

**Comment by KrainskiL** [comment]
Linux shasum examples can be provided in comments (as in commit) or as main example?

File: utils/graph_sampler.jl (hunk 7 above)
Comment (KrainskiL) [reply utils/graph_sampler.jl:4]: I'm not sure what is missing - `seed` is specified in Usage, Example and is processed in line 23 below

File: utils/abcd_sampler.jl (hunk 4 above)
Comment (bkamins) [reply utils/abcd_sampler.jl:39]: if the error is informative then no need to add it. If the error is messy then please add

File: utils/graph_sampler.jl (hunk 7 above)
Comment (bkamins) [reply utils/graph_sampler.jl:4]: I missed 23

**Comment by bkamins** [comment]
> Linux shasum examples can be provided in comments (as in commit) or as main example?

As you prefer. Let us just avoid Win10 users running it :smile:

File: utils/abcd_sampler.jl (hunk 4 above)
Comment (KrainskiL) [reply utils/abcd_sampler.jl:39]: Yes, it's straightforward 

> "when ξ is provided local model is not allowed"

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**